### PR TITLE
CRI: Add devices to ContainerConfig

### PR DESCRIPTION
**Pull request by feiskyer**
This PR adds devices to ContainerConfig and adds experimental GPU support.

cc/ @yujuhong @Hui-Zhi @vishh @kubernetes/sig-node

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35597)

### CODE REVIEW

**Comment by yujuhong** [comment]
I thought we decided to use annotations for experimental features such as this for now.

**Comment by feiskyer** [comment]
> I thought we decided to use annotations for experimental features such as this for now.

@yujuhong Although GPU is an alpha feature, it is not set by annotations. It is part of resource limits.

**Comment by yujuhong** [comment]
@vishh since you reviewed the original PR for adding devices, do you want to take a look? Feel free to remove yourself if you are too busy.

/cc @kubernetes/sig-rktnetes @euank @yifan-gu @mrunalp

**Comment by feiskyer** [comment]
Rebased.

**Comment by Hui-Zhi** [comment]
I think for dedicated GPU devices, it is OK. But if we want to support GPU cores/memory schedule, we need to change it.

**Comment by feiskyer** [comment]
> I think for dedicated GPU devices, it is OK. But if we want to support GPU cores/memory schedule, we need to change it.

@Hui-Zhi It seems GPU cores/memory will not be supported in near future (at least not in 1.5). There is a few PR/proposal for extending GPU features, and they are already covered here.

**Comment by Hui-Zhi** [comment]
@feiskyer Sure, and so far this PR LGTM.

**Comment by yujuhong** [comment]
Looks reasonable to me. Anyone from rkt wants to take a look? @euank @yifan-gu

**Comment by mrunalp** [comment]
Looks good.

**Comment by euank** [comment]
The user can't specify the permissions in any way, can they? If they're just omitted, it would be assumed that they'd match the host permissions, which is what we'd want anyways, right?

**Comment by mrunalp** [comment]
@euank Good point. How about defaulting to match the host but still allowing overrides? We could have a UID, GID and FileMode like in the OCI runtime spec. https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L336

**Comment by euank** [comment]
@mrunalp I'd rather add things when they're needed for the sake of simplicity and ease of correct implementation.

**Comment by mrunalp** [comment]
@euank Actually, looking closer, the permissions here look like the device cgroup settings. And we would want stricter controls in the container compared to the host. So, I think that these should be added. (Defined in OCI runtime spec here https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L355)

**Comment by feiskyer** [comment]
> The user can't specify the permissions in any way, can they? If they're just omitted, it would be assumed that they'd match the host permissions, which is what we'd want anyways, right?

@euank Yes for current GPU devices, it is always 'mrw'. But we may add other devices in the future which are readonly, so it's better to keep the permissions.

**Comment by yujuhong** [comment]
@euank I think the group device permission may be used soon for other devices. I am ok keeping this PR as it is. WDYT?

**Comment by feiskyer** [comment]
@yujuhong @euank Rebased. PTAL.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 1]
```diff
@@ -428,6 +428,19 @@ message ContainerMetadata {
...
 }
 
+// Device specifies a host device to mount into a container.
+message Device {
+    // The path of the device within the container.
+    optional string container_path = 1;
+    // The path of the device on the host.
+    optional string host_path = 2;
+    // Cgroups permissions of the device, candidates are one or more of
+    // * r — allows container to read from the specified device.
```
Comment (euank) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: nit: An ascii `-` is better for grepping.

**Comment by euank** [comment]
@yujuhong this doesn't capture the actual complexity of device node permissions vs device cgroup entries though.

As @mrunalp linked above, the permissions are used for creating a device cgroup entry which happens to reference the same path as the bindmount, but can exist independently of the bindmount (a use-case for that is if the container wishes to mknod `/dev/net/tun`; not use the hosts.. in that case, it wants a cgroup permission to do it with no mount)

All the runtimes already do the right thing for gpu devices without adding in this permission field yet. Because of the additional complexity that brings, I'm still in favor of punting on defining that field.

I don't feel strongly enough to block this though, so if you feel strongly this is the right way to define it and that we should later change or remove this if we wish to have independent device cgroup entries, that's fine by me.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: ack. thanks.

**Comment by feiskyer** [comment]
> if the container wishes to mknod /dev/net/tun; not use the hosts.. in that case, it wants a cgroup permission to do it with no mount

I don't think the case is covered by devices in this PR:

- [Device](https://github.com/kubernetes/kubernetes/pull/35597/files#diff-ae4e12492f911b555e0870ec9962c634R431) specifies a host device to mount into a container. It assumes the device of host_path already exists.  
- On the other hand,  current`Device` struct doesn't provides enough information for containers to create a new device, e.g. it doesn't provides device type or major/minor numbers. 

> All the runtimes already do the right thing for gpu devices without adding in this permission field yet. Because of the additional complexity that brings, I'm still in favor of punting on defining that field.

Only docker runtime supports GPU today. other runtimes (e.g. rkt/cri-o) may want to support GPU too, but I'm not sure whether the permissions are set correctly or not.

> we should later change or remove this if we wish to have independent device cgroup entries, 

Does `independent device cgroup` mean to cover newly-created container-internal devices (e.g. by `/dev/tun/net` ) here? Its permission is controlled by both capabilites and [device cgroups](https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt). And it is out the scope of this PR. Maybe open another issue for this?

**Comment by feiskyer** [comment]
@k8s-bot unit test this

**Comment by yujuhong** [comment]
@euank the use case you mentioned is not really in the scope of the "device" usage in kubelet. If we ever want to expand the usage, it'd be reasonable to change the API to reflect that. For now, I think this PR is fine :-)

**Comment by yujuhong** [comment]
LGTM

**Comment by feiskyer** [comment]
@yujuhong Rebased. PTAL.

**Comment by feiskyer** [comment]
@yujuhong Rebased. PTAL.

**Pull request merged**